src/modules/bulletproofs/inner_product_impl.h

```diff
@@ -0,0 +1,848 @@
+    secp256k1_scalar yinv;
+    unsigned char commit[32];
+    secp256k1_bulletproof_vfy_callback *rangeproof_cb;
+    void *rangeproof_cb_data;
+    size_t n_extra_rangeproof_points;
+} secp256k1_bulletproof_innerproduct_context;
+
+/* used internally */
+typedef struct {
+    const secp256k1_bulletproof_innerproduct_context *proof;
+    secp256k1_scalar abinv[IP_AB_SCALARS];
+    secp256k1_scalar xsq[SECP256K1_BULLETPROOF_MAX_DEPTH + 1];
+    secp256k1_scalar xsqinv[SECP256K1_BULLETPROOF_MAX_DEPTH + 1];
+    secp256k1_scalar xsqinvy[SECP256K1_BULLETPROOF_MAX_DEPTH + 1];
+    secp256k1_scalar xcache[SECP256K1_BULLETPROOF_MAX_DEPTH + 1];
+    secp256k1_scalar xsqinv_mask;
+    const unsigned char *serialized_lr;
+} secp256k1_bulletproof_innerproduct_vfy_data;
+
+/* used by callers to modify the multiexp */
+typedef struct {
+    size_t n_proofs;
+    secp256k1_scalar p_offs;
+    const secp256k1_ge *g;
+    const secp256k1_ge *geng;
+    const secp256k1_ge *genh;
+    size_t vec_len;
+    size_t lg_vec_len;
+    int shared_g;
+    secp256k1_scalar *randomizer;
+    secp256k1_bulletproof_innerproduct_vfy_data *proof;
+} secp256k1_bulletproof_innerproduct_vfy_ecmult_context;
+
+size_t secp256k1_bulletproof_innerproduct_proof_length(size_t n) {
+    if (n < IP_AB_SCALARS / 2) {
+        return 32 * (1 + 2 * n);
+    } else {
+        size_t bit_count = POPCOUNT(n);
+        size_t log = secp256k1_floor_lg(2 * n / IP_AB_SCALARS);
+        return 32 * (1 + 2 * (bit_count - 1 + log) + IP_AB_SCALARS) + (2*log + 7) / 8;
+    }
+}
+
+/* Our ecmult_multi function takes `(c - a*b)*x` directly and multiplies this by `G`. For every other
+ * (scalar, point) pair it calls the following callback function, which takes an index and outputs a
+ * pair. The function therefore has three regimes:
+ *
+ * For the first `n` invocations, it returns `(s'_i, G_i)` for `i` from 1 to `n`.
+ * For the next `n` invocations, it returns `(s_i, H_i)` for `i` from 1 to `n`.
+ * For the next `2*log2(n)` invocations it returns `(x_i^-2, L_i)` and `(x_i^2, R_i)`,
+ *   alternating between the two choices, for `i` from 1 to `log2(n)`.
+ *
+ * For the remaining invocations it passes through to another callback, `rangeproof_cb_data` which
+ * computes `P`. The reason for this is that in practice `P` is usually defined by another multiexp
+ * rather than being a known point, and it is more efficient to compute one exponentiation.
+ *
+ * Inline we refer to the first `2n` coefficients as `s_i` for `i` from 0 to `2n-1`, since that
+ * is the more convenient indexing. In particular we describe (a) how the indices `j` and `k`,
+ * from the big comment block above, are chosen; and (b) when/how each `x_k^-2` is computed.
+ */
+static int secp256k1_bulletproof_innerproduct_vfy_ecmult_callback(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *data) {
+    secp256k1_bulletproof_innerproduct_vfy_ecmult_context *ctx = (secp256k1_bulletproof_innerproduct_vfy_ecmult_context *) data;
```

Comment:
I think it is guaranteed by the caller and nothing really bad happens but I think we should check that `ctx->vec_len` is not 0 here.

Reply:
It's fine if it is. In fact the inner product unit tests check this case.